urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/laazotea/indi/driver/INDIFocuserDriver.java

```diff
@@ -0,0 +1,373 @@
+   * <code>FOCUS_SPEED_VALUE</code> element.
+   *
+   * @return The current speed value
+   */
+  protected int getCurrentSpeed() {
+    if (focusSpeedValueE != null) {
+      int speed = focusSpeedValueE.getValue().intValue();
+
+      return speed;
+    }
+
+    return -1;
+  }
+
+  /**
+   * Gets the desired absolute position (which may not have been reached by the
+   * focuser).
+   *
+   * @return The desired absolute position
+   */
+  protected int getDesiredAbsPosition() {
+    return desiredAbsPosition;
+  }
+
+  /**
+   * Must be called by drivers when the final position for the focuser has been
+   * reached.
+   */
+  protected void finalPositionReached() {
+    absFocusPositionP.setState(PropertyStates.OK);
+
+    try {
+      updateProperty(absFocusPositionP);
+    } catch (INDIException e) {
+    }
+  }
+
+  /**
+   * Must be called by drivers when a new speed has been set.
+   */
+  protected void desiredSpeedSet() {
+    focusSpeedP.setState(PropertyStates.OK);
+
+    try {
+      updateProperty(focusSpeedP);
+    } catch (INDIException e) {
+    }
+  }
+
+  /**
+   * Must be called by drivers when the focuser stops (only when it has been
+   * asked to stop)
+   */
+  protected void stopped() {
+    stopFocusingP.setState(PropertyStates.OK);
+
+    try {
+      updateProperty(stopFocusingP);
+    } catch (INDIException e) {
+    }
+  }
+
+  /**
+   * Should be called by the drivers when the focuser its moving. It can be
+   * called with any frequency, but a less than one second is preferred to
+   * notify the clients of the movement of the focuser.
+   *
+   * @param currentPos The current position of the focuser.
+   */
+  protected void positionChanged(int currentPos) {
+    focusAbsolutePositionE.setValue("" + currentPos);
+
+    try {
+      updateProperty(absFocusPositionP);
+    } catch (INDIException e) {
+    }
+  }
+  
+  /**
+   * Should be called by the drivers when the focuser speed changes (if for
+   * example the device has a potentiometer to control the speed). 
+   *
+   * @param currentSpeed The current speed of the focuser.
+   */
+  protected void speedChanged(int currentSpeed) {
+    focusSpeedValueE.setValue("" + currentSpeed);
+
+    try {
+      updateProperty(focusSpeedP);
+    } catch (INDIException e) {
+    }
+  }
+
+  @Override
+  public void processNewSwitchValue(INDISwitchProperty property, Date timestamp, INDISwitchElementAndValue[] elementsAndValues) {
+    if (property == stopFocusingP) {
+      stopFocusingP.setState(PropertyStates.BUSY);
+      stopFocusingP.setStatus(SwitchStatus.OFF);
+      try {
+        updateProperty(stopFocusingP);
+      } catch (INDIException e) {
+      }
+
+      stopHasBeenRequested();
+    }
+  }
+
+  @Override
+  public void processNewNumberValue(INDINumberProperty property, Date timestamp, INDINumberElementAndValue[] elementsAndValues) {
+    if (property == absFocusPositionP) {
+      int newVal = elementsAndValues[0].getValue().intValue();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;elementsAndValues&#39; (index: 2 | type: INDINumberElementAndValue[). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;property&#39;. The same parameter type is validated here for example: urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/farom/astroiddriver/INDIAstroidDriver.java:387. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).